design/build-clusters.md

```diff
@@ -0,0 +1,176 @@
+
+#### Example: 
+
+1. Queue worker invokes `getJobs/:buildCluster` api 
+
+2. Create JWT token with below information and pass it alongwith `getJobs/:buildCluster` api header.
+
+    1. Payload: { buildCluster: colo1 }
+    2. Secret: privatekey
+    3. SignOption: algorithm, expiry, etc.
+
+3. Scheduler service 
+    1. gets buildCluster name from the getJobs api parameter
+    2. gets respective buildCluster public key based on the buildCluster name from #1 
+    3. Verifies JWT sent by queue worker
+    4. if legit, then decodes the JWT content 
+    5. pulls message from the respective queue
+    6. responds back build info to be processed to queue worker.   
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```   
+ 
+### New table for build cluster details
+Table: `buildClusters`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | no | no | no | |
+| `scmContext` | text(200) | no | no | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `authKey` | text(100) | no | no | yes | *environment variable name of buildCluster publicKey. Note: every buildCluster publicKey will have unique environment variable name* |
+| `signOption` | text(100) | no | no | yes | *environment variable name of buildCluster signOption. Note: every buildCluster publicKey will have unique environment variable name* |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+
+Unique constraint: `name + isActive` 
+
+#### Sample record
+
+| id | name | scmContext | scmOrganizations | isActive | authKey |  signOption | managedBy 
+| --- | --- | --- | --- | --- | --- | --- | --- | 
+| 1 | gq1 | github:git.ouroath.com | null | 0 | *gq1_sdpublickey* | *gq1_sign* | screwdriver
+| 2 | bf1 | github:git.ouroath.com | null | 1 | *bf1_sdpublickey* | *bf1_sign* | screwdriver
+| 3 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 1 | *identity_publickey* | *identity_sign* | external
+| 4 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 0 | *identity_publickey* | *identity_sign* | external
+| 5 | iOS | github:git.ouroath.com | iOS_org1, iOS_org2 | 1 | *iOS_publickey* | *iOS_sign* | external
+
+### Cache active *buildClusters* in Redis 
+    Scheduler service will be using this cache to get active buildClusters and respective queues.
```

Comment:
why do we need this ? this details can be stored in DB and cache server can store them in memory when the service  boots up.